include/zephyr/sys/arch_interface.h

```diff
@@ -307,6 +307,16 @@ int arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
 			     void (*routine)(const void *parameter),
 			     const void *parameter, uint32_t flags);
 
+/**
+ * Arch-specific hook to set an interrupt's priority
+ *
+ * @param irq IRQ line number
+ * @param priority Interrupt priority
+ * @param flags Arch-specific IRQ configuration flag
+ */
+void arch_irq_priority_set(unsigned int irq, unsigned int priority,
```

Comment:
Hi, I'm wondering about the rational of such API? Is there any real scenario when we need to set priority not once on the init but in the runtime?

If we do want to make this API something standard we should think more about the corner cases.

For example, on ARC SMP target that operation isn't atomic (after been properly implemented) as we need to set interrupt priorities on each core private interrupt controller. That basically means we need to do inter-core interrupts to make this setup. 

So doing that such interrupt priority setup is not the thing which is safe when the interrupt line is enabled and interrupt may arrive.

IMO, at least we should add retirement that interrupt line should be disabled when we do dynamic priority setup.

Reply:
Hi, thanks for your comment. I think there are use-cases for priority changing during the runtime of an application. Priority is set at runtime anyway (even if during init) so this change just allows it to be set by the application at any point. Also, `irq_connect_dynamic()` already allows runtime interrupt priority setting, this new API does the same but without configuring the ISR.

You make a good point that this operation could be unsafe and taking care such as disabling the interrupt line while priority is changed would be sensible. I think this should be handled by each architecture's implementation - I had assumed that as `irq_connect_dynamic()` already configures interrupt priority at runtime that architectures should already have safe implementations. Note that this PR doesn't add any architecture-specific implementations, it just uses already existing ones.

Reply:
> I think there are use-cases for priority changing during the runtime of an application.

So what is the reason for adding the API if we don't even have an idea how we can benefit from it? :)

>> so this change just allows it to be set by the application at any point.
>
>> Note that this PR doesn't add any architecture-specific implementations, it just uses already existing ones.

That is completely not true! Currently 'dynamic interrupt priority setup' has multiple restrictions which we break by this new API.

Currently the dynamic interrupt setup may be done only **once** during the system execution as `irq_connect_dynamic` may be called only once per interrupt line. It's clearly specified in the documentation, see:  

```
Removing or re-configuring a dynamic interrupt is currently unsupported.
```

Additional conclusion from this is that dynamic priority setup can be done only on disabled interrupt line - as we can't enable line if we don't instantiate IRQ handler (with `irq_connect_dynamic`).

And there is deep rational behind all that limitations - we can't just change interrupt priority on the enabled line without IRQ handling code overhaul (or at least verify) on all architectures - as currently we don't expect that we may enter the interrupt with one priority, but exit with it's priority changed. Please don't forget that Zephyr supports multi-core systems and it's perfectly valid corner case! (when we process IRQ on one core and change this IRQ priority from another core).

Moreover such scenario may not be supported by some interrupt controllers HW itself.

---

In conclusion. We try to introduce the API:

- we don't have clear vision about how and why it can / should be used
- which breaks current assumptions (and we don't verify if it is actually valid without such assumptions)
- we don't sure if it can properly implemented on all architectures

I'd propose to avoid that.

Reply:
> So what is the reason for adding the API if we don't even have an idea how we can benefit from it? :)

I wouldn't have requested the feature if I had no use for it :) My bad, I should have explained my use-cases. Consider any application that has multiple operating modes/states, and between these different modes/states the relative priorities of interrupt-driven functionality change. Also, this would overcome a limitation of setting interrupt priority statically through devicetree where (please correct me if I'm wrong) to set the interrupt priority of a device you must also set the IRQ line even though this is already set and will not be changed.

> Currently the dynamic interrupt setup may be done only once during the system execution as `irq_connect_dynamic` may be called only once per interrupt line.

> Additional conclusion from this is that dynamic priority setup can be done only on disabled interrupt line - as we can't enable line if we don't instantiate IRQ handler (with `irq_connect_dynamic`).

Okay, yes, I hadn't appreciated that the use of `irq_connect_dynamic` relied upon the interrupt line already being disabled. Would there be a problem with disabling the interrupt line prior to changing the interrupt priority to make this a safe operation? That wouldn't be hard to add, but if that's not sufficient then yes this feature does become more problematic.

> we don't sure if it can properly implemented on all architectures

I don't think that not being supported on all architectures/HW should stop the feature being introduced, interrupt priority is not supported on all archs/HW and of course that is supported by Zephyr.

Reply:
> Would there be a problem with disabling the interrupt line prior to changing the interrupt priority to make this a safe operation?

Depends on the interrupt controller. I have x86 IOAPIC in mind but I guess this might be the case elsewhere too. You would need to have a migration procedure: where old setup would need a proper closure (make sure pending irq are handled etc...), disabling, then de-allocate any related object, reallocate, enabling... 

Imo, if such feature is required, it will be nice to see it properly implemented for all relevant arch instead of having dummy functions doing nothing.